lib/find-python.js

```diff
@@ -1,13 +1,26 @@
 'use strict'
 
-const path = require('path')
 const log = require('npmlog')
 const semver = require('semver')
 const cp = require('child_process')
 const extend = require('util')._extend // eslint-disable-line
 const win = process.platform === 'win32'
 const logWithPrefix = require('./util').logWithPrefix
 
+const systemDrive = process.env.SystemDrive || 'C:'
+const username = process.env.USERNAME || process.env.USER
```

Comment:
I think this should fall back to assigning some string to `username`. Even something not very useful like `''` or `'User'`.

(It's possible both `process.env.USERNAME` and `process.env.USER` are `undefined`. Having a not very useful path to check is better than causing a `ReferenceError`.)

Reply:
I actually like the error message better than unsuccessfully trying to sweep the problem under the carpet.  The message makes it clear what the problem is.

Please do not worry about the Python 4 case.  We will cross that bridge later.

Reply:
Thinking a bit more...

If username is undefined and we can find a globally installed acceptable Python then we should use that without complaint.  If we fail to find an acceptable Python and username is undefined then we should raise the error.

Reply:
You can also get the username via `os.userInfo()`.

Reply:
Thanks for that suggestion @cjihrig. `os.userInfo()` should always return a username string. I committed that as a fallback. I think this can reasonably be expected to prevent `Uncaught ReferenceError: username is not defined`.

Reply:
Note: There was apparently no risk of `Uncaught ReferenceError` after all.

<details>

If you write `.someNotDefinedThing` off of an object or array, you get `undefined` as the result, but no `ReferenceError`. Since `process.env` returns an object, doing `process.env.THIS_IS_NOT_DEFINED` does not produce a `ReferenceError`.

For example:

```
console.log(`Here is my undefined var: ${{}.some_undefined_var}`)
```

Outputs the string:

```
Here is my undefined var: undefined
```

So we would construct a Python path resembling this: `C:\\Users\\undefined\\AppData\\Local\\Programs\\Python\\Python39\\python.exe`

</details>

I'm happy with the `os.userInfo()` fallback. It makes it pretty much guaranteed we can find the username. This is just an interesting finding from my tests.

Reply:
Related to this, still an issue when the username is not defined, only the user ID e.g. that is common in 'non-root' containers for OpenShift https://docs.openshift.com/container-platform/3.3/creating_images/guidelines.html#openshift-container-platform-specific-guidelines.

```
I have no name!@52545ce360b8$ node-gyp --help
os.js:272
    throw new ERR_SYSTEM_ERROR(ctx);
    ^

SystemError [ERR_SYSTEM_ERROR]: A system error occurred: uv_os_get_passwd returned ENOENT (no such file or directory)
    at Object.userInfo (os.js:272:11)
    at Object.<anonymous> (/opt/bitnami/node/lib/node_modules/node-gyp/lib/find-python.js:11:76)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/opt/bitnami/node/lib/node_modules/node-gyp/lib/configure.js:11:18)
    at Module._compile (internal/modules/cjs/loader.js:999:30) {
  code: 'ERR_SYSTEM_ERROR',
  info: {
    errno: -2,
    code: 'ENOENT',
    message: 'no such file or directory',
    syscall: 'uv_os_get_passwd'
  },
  errno: [Getter/Setter],
  syscall: [Getter/Setter]
}
```

Python is in the path so a good option could be to add a fallback or not return an error for the line

```
const username = process.env.USERNAME || process.env.USER || require('os').userInfo().username
```